source/bsd/kqueue_event_loop.c

```diff
@@ -297,7 +297,8 @@ static void s_destroy(struct aws_event_loop *event_loop) {
         return;
     }
     /* setting this so that canceled tasks don't blow up when asking if they're on the event-loop thread. */
-    aws_atomic_store_int(&impl->thread_id, (size_t)aws_thread_current_thread_id());
+    aws_thread_id current_thread_id = aws_thread_current_thread_id();
+    aws_atomic_store_ptr(&impl->thread_id, &current_thread_id);
```

Comment:
This seems dangerous, storing the pointer to a stack variable somewhere

Reply:
In this case it is safe because the event loop is being destroyed. The pointer will be free'd by `aws_mem_release(event_loop->alloc, impl)` after the task scheduler has been flushed.

The alternatives we discussed:
  - Use an `AWS_THREAD_LOCAL aws_thread_id` variable
  - Reset pointer to NULL after task scheduler has been flushed

Reply:
It's not safe due to how the atomic is accessed in `s_is_event_thread()`

In that function, it loads the atomic (a pointer), then accesses the value that it's pointing to. If, in the moment between loading the atomic and following the pointer, the thread is deleted, then we will access invalid memory when we follow the pointer.

Reply:
Ah yes, I see the problem. This means thread-local storage won't work either since the race is possible there too.

Reply:
Hi both, the last change no longer takes a pointer to a stack-local variable. Instead, we re-use the variable `event_loop->impl.thread.thread_id`. At this point of the execution, the event loop thread has been joined to the current thread (by `wait_for_stop_completion()`) so we can reuse the var.

Regarding the potential race if one thread is calling `destroy()` concurrently with another calling `is_on_callers_thread()`. I think this is a user error. If thread T is calling `destroy()` on an event loop then all other threads should have given up their right to call event loop functions before T calls destroy. Otherwise, there is the danger that these calls will encounter a free'd event loop object.

What if we add the following text to the API call `aws_event_loop_destroy`?

> If the event loop is shared by multiple threads then destroy must be called by exactly one thread. All other threads must ensure their API calls to the event loop happen-before the call to destroy.

Reply:
Can we check for this at runtime and throw an error?

Reply:
@danielsn We could set an atomic flag at the top of destroy() and then check against that in all API calls. This wouldn't catch all misuses, but could be useful in debug builds.

Resolving the original request.